src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioDocumentTracker.cs

```diff
@@ -174,23 +157,29 @@ public override ITextView GetFocusedTextView()
             return null;
         }
 
-        public void Subscribe()
+        public void Subscribe(WorkspaceState workspaceState)
         {
+            if (workspaceState == null)
+            {
+                throw new ArgumentNullException(nameof(workspaceState));
+            }
+
             _foregroundDispatcher.AssertForegroundThread();
 
             if (_subscribeCount++ > 0)
             {
                 return;
             }
 
-            _projectSnapshot = _projectManager.GetOrCreateProject(_projectPath);
+            _workspaceState = workspaceState;
+            _projectSnapshot = _workspaceState.ProjectSnapshotManager.GetOrCreateProject(_projectPath);
```

Comment:
So, if I understand it correctly, `WorkspaceState` is acting a lot like `WorkspaceRegistration` from Roslyn's code right?

Reply:
No. It's just a helper object that extracts services from the Workspace so you don't need to do it at a later date. Before this most of the touched types in this PR were constructed with a workspace available so they were constructed already having access to all the various Workspace services. Since we can't have the services on construction we wait to do our initialization process until we've been handed the services needed to fully subscribe to the workspace.